src/aics-image-viewer/components/useVolume.ts

```diff
@@ -0,0 +1,409 @@
+import { LoadSpec, RawArrayLoaderOptions, View3d, Volume, VolumeLoaderContext } from "@aics/vole-core";
+import { useCallback, useContext, useEffect, useMemo, useRef, useState } from "react";
+import { Box3, Vector3 } from "three";
+
+import {
+  AXIS_TO_LOADER_PRIORITY,
+  CACHE_MAX_SIZE,
+  DTYPE_RANGE,
+  getDefaultChannelColor,
+  QUEUE_MAX_LOW_PRIORITY_SIZE,
+  QUEUE_MAX_SIZE,
+} from "../shared/constants";
+import { ViewMode } from "../shared/enums";
+import { AxisName } from "../shared/types";
+import {
+  controlPointsToRamp,
+  initializeLut,
+  rampToControlPoints,
+  remapControlPointsForChannel,
+} from "../shared/utils/controlPointsToLut";
+import { useConstructor, useRefWithSetter } from "../shared/utils/hooks";
+import PlayControls from "../shared/utils/playControls";
+import SceneStore from "../shared/utils/sceneStore";
+import { ChannelGrouping, getDisplayName, makeChannelIndexGrouping } from "../shared/utils/viewerChannelSettings";
+import { initializeOneChannelSetting } from "../shared/utils/viewerState";
+import { ChannelState } from "./ViewerStateProvider/types";
+
+import { ViewerStateContext } from "./ViewerStateProvider";
+
+export type UseVolumeOptions = {
+  onChannelLoaded?: (image: Volume, channelIndex: number, channelSettings: ChannelState) => void;
+  onError?: (error: unknown) => void;
+  maskChannelName?: string;
+};
+
+export const enum ImageLoadStatus {
+  REQUESTED,
+  LOADING,
+  LOADED,
+  ERROR,
+}
+
+// Used by `channelVersions` (see below)
+const CHANNEL_INITIAL_LOAD = -1;
+const CHANNEL_RELOAD = 0;
+
+export type ReactiveVolume = {
+  image: Volume | null;
+  /**
+   * Indicates the load status of each channel:
+   *
+   * - `-1` indicates the channel has not yet loaded and will get some extra initialization (e.g. LUTs) when it loads.
+   * - `0` indicates the channel has been loaded once, but is currently waiting for new data.
+   * - `1` or greater indicates the channel is loaded. Note this is not exactly `1` to handle multiple simultaneous
+   *   loads: if multiple loads are issued before the first one completes, _incrementing_ rather than _setting_ the
+   *   version number means we react to each load when it completes, rather than just the first.
+   */
+  channelVersions: number[];
```

Comment:
I wish this was an array of objects containing a load status Enum + a version number...

Reply:
I definitely would prefer better typing, but what I like about the bare numbers is that they capture the possible states more tightly, where in the enum-plus-version approach the version field is unused in all but the "loaded" state. Outside of this hook these values are only ever used as a "key" to determine when to run channel update effects, and these semantics are barely used (there's one `> 0` check for "is this channel loaded"). This could be a union with a string-typed enum (`"init"` and `"reload"`?)?